In [1]:
import numpy as np
import pandas as pd

In [2]:
reldir = 'C:\\Users\\shefa\\OneDrive\\CS6220\\msd.recsys\\datasets\\'

In [3]:
standardized_df = pd.read_pickle(reldir + 'combined.csv', compression='zip')

In [4]:
master_df = pd.read_pickle(reldir + 'songdata.csv', compression='zip')

In [5]:
standardized_df.tail()

user                song  count  \
772656  3a2d1642c26b16b178a43bead38ea1fe6883e543  SOZSHWZ12A8C1452B5      1   
772657  a92dcaddbdf2e46c40488823431d980614ab9c66  SOKTKTM12AB0189837      1   
772658  6b9df737dae978d4b8a2099ffa6a501a3c30b091  SOMLFVN12A8C136135      1   
772659  6b9df737dae978d4b8a2099ffa6a501a3c30b091  SOYUIMG12A8C135784      1   
772660  6428bbc86c59bd5d1d289935056b57acd0e050ad  SOCDLVW12AB018371F      5   

       analysis_songs_track_id metadata_songs_artist_id  \
772656      TRATKXV128F92D0480       ARKSLFG1187B99DADD   
772657      TRAIZIR128F934B3A4       ARQFKVO1187B9B1777   
772658      TRAGCSE128F425FE4B       ARTK5LU1187B9A654C   
772659      TRAVXZP128F425C677       ARTK5LU1187B9A654C   
772660      TRASFAV128F93428A4       ARV1JVD1187B9AD195   

       metadata_songs_song_id  analysis_songs_key  \
772656     SOZSHWZ12A8C1452B5                   9   
772657     SOKTKTM12AB0189837                   1   
772658     SOMLFVN12A8C136135                   7   
772659     SOYUIMG12A8C135784                   1   
772660     SOCDLVW12AB018371F                   0   

        analysis_songs_key_confidence  analysis_songs_mode  \
772656                          0.868                    1   
772657                          0.491                    1   
772658                          0.369                    1   
772659                          0.156                    0   
772660                          0.669                    1   

        analysis_songs_mode_confidence  ...  tempo_binned  \
772656                           0.772  ...             9   
772657                           0.357  ...             7   
772658                           0.406  ...             3   
772659                           0.150  ...             8   
772660                           0.696  ...             8   

        analysis_songs_duration  metadata_songs_artist_familiarity  \
772656                -0.797657                          -0.166310   
772657                -0.962968                          -0.396293   
772658                 1.075431                          -0.264920   
772659                 2.284152                          -0.264920   
772660                -0.706801                           0.485700   

        metadata_songs_artist_hotttnesss  metadata_songs_song_hotttnesss  \
772656                         -0.077048                       -4.007280   
772657                         -0.348441                       -0.230387   
772658                         -0.365274                       -0.899601   
772659                         -0.365274                       -0.941122   
772660                          0.482326                        0.516715   

                                    metadata_artist_terms  \
772656  b'power pop'|b'jangle pop'|b'garage rock'|b'ne...   
772657  b'shock rock'|b'hard rock'|b'punk'|b'united st...   
772658  b'chamber jazz'|b'chill-out'|b'easy listening'...   
772659  b'chamber jazz'|b'chill-out'|b'easy listening'...   
772660  b'modern rock'|b'singer-songwriter'|b'folk'|b'...   

        metadata_songs_artist_name  \
772656                    The Keys   
772657     Dave Brockie Experience   
772658              Michael Gettel   
772659              Michael Gettel   
772660       Loudon Wainwright III   

                                   metadata_songs_release  \
772656                                           The Keys   
772657                                Songs For The Wrong   
772658  San Juan Suite/San Juan Suite II: Narada Classics   
772659                                     Places In Time   
772660   High Wide & Handsome - The Charlie Poole Project   

                metadata_songs_title  \
772656  From Tense To Loose To Slack   
772657                Slips Of Paper   
772658                  Safe Passage   
772659                     River Run   
772660                        Rowena   

                                musicbrainz_artist_mbtags  
772656       

In [6]:
standardized_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 772661 entries, 0 to 772660
Data columns (total 27 columns):
 #   Column                                     Non-Null Count   Dtype  
---  ------                                     --------------   -----  
 0   user                                       772661 non-null  object 
 1   song                                       772661 non-null  object 
 2   count                                      772661 non-null  int64  
 3   analysis_songs_track_id                    772661 non-null  object 
 4   metadata_songs_artist_id                   772661 non-null  object 
 5   metadata_songs_song_id                     772661 non-null  object 
 6   analysis_songs_key                         772661 non-null  int64  
 7   analysis_songs_key_confidence              772661 non-null  float64
 8   analysis_songs_mode                        772661 non-null  int64  
 9   analysis_songs_mode_confidence             772661 non-null  float64
 10  analysis

https://jessesw.com/Rec-System/

https://towardsdatascience.com/building-a-collaborative-filtering-recommender-system-with-clickstream-data-dffc86c8c65

In [7]:
master_df = standardized_df[standardized_df.groupby('user').user.transform('count')>4].copy()
#master_df.loc[master_df['user']=="a92dcaddbdf2e46c40488823431d980614ab9c66"]
master_df.tail()

user                song  count  \
772636  da9ccab0fcdc8240cf4e352f2cc530d0c69dd8e1  SOVSNJX12AB0189848      1   
772638  bee388dcd9da37c9b40d7f43ecf79c440a49d608  SOCCQJE12AB018722F      5   
772657  a92dcaddbdf2e46c40488823431d980614ab9c66  SOKTKTM12AB0189837      1   
772658  6b9df737dae978d4b8a2099ffa6a501a3c30b091  SOMLFVN12A8C136135      1   
772659  6b9df737dae978d4b8a2099ffa6a501a3c30b091  SOYUIMG12A8C135784      1   

       analysis_songs_track_id metadata_songs_artist_id  \
772636      TRBACAA12903D0575F       ARKL75U1187FB5B8B6   
772638      TRAULGT12903D10B94       AR4GAFY1187B9B639D   
772657      TRAIZIR128F934B3A4       ARQFKVO1187B9B1777   
772658      TRAGCSE128F425FE4B       ARTK5LU1187B9A654C   
772659      TRAVXZP128F425C677       ARTK5LU1187B9A654C   

       metadata_songs_song_id  analysis_songs_key  \
772636     SOVSNJX12AB0189848                   9   
772638     SOCCQJE12AB018722F                  11   
772657     SOKTKTM12AB0189837                   1   
772658     SOMLFVN12A8C136135                   7   
772659     SOYUIMG12A8C135784                   1   

        analysis_songs_key_confidence  analysis_songs_mode  \
772636                          0.571                    1   
772638                          0.246                    1   
772657                          0.491                    1   
772658                          0.369                    1   
772659                          0.156                    0   

        analysis_songs_mode_confidence  ...  tempo_binned  \
772636                           0.520  ...             7   
772638                           0.622  ...             0   
772657                           0.357  ...             7   
772658                           0.406  ...             3   
772659                           0.150  ...             8   

        analysis_songs_duration  metadata_songs_artist_familiarity  \
772636                -0.069509                           0.268770   
772638                -0.365246                           0.723507   
772657                -0.962968                          -0.396293   
772658                 1.075431                          -0.264920   
772659                 2.284152                          -0.264920   

        metadata_songs_artist_hotttnesss  metadata_songs_song_hotttnesss  \
772636                          0.196163                        0.145691   
772638                          1.256276                        0.602844   
772657                         -0.348441                       -0.230387   
772658                         -0.365274                       -0.899601   
772659                         -0.365274                       -0.941122   

                                    metadata_artist_terms  \
772636  b'roots rock'|b'heartland rock'|b'pop rock'|b"...   
772638  b'blues-rock'|b'blues'|b'jazz'|b'guitar'|b'cou...   
772657  b'shock rock'|b'hard rock'|b'punk'|b'united st...   
772658  b'chamber jazz'|b'chill-out'|b'easy listening'...   
772659  b'chamber jazz'|b'chill-out'|b'easy listening'...   

        metadata_songs_artist_name  \
772636                     BoDeans   
772638                     BB King   
772657     Dave Brockie Experience   
772658              Michael Gettel   
772659              Michael Gettel   

                                   metadata_songs_release  \
772636                                      Mr. Sad Clown   
772638                            The Road from Itta Bena   
772657                                Songs For The Wrong   
772658  San Juan Suite/San Juan Suite II: Narada Classics   
772659                                     Places In Time   

             metadata_songs_title  \
772636                  Back Then   
772638  She Don't Move Me No More   
772657             Slips Of Paper   
772658               Safe Passage   
772659                  River Run   

                                musicbrainz_artist_mbtags  
772636                         

## Attempt 1 Using jessesw.com

In [8]:
## Check that all counts in dataset are 1 or greater
standardized_df.describe()

count  analysis_songs_key  analysis_songs_key_confidence  \
count  772661.000000       772661.000000                  772661.000000   
mean        2.684340            5.673522                       0.432972   
std         5.454646            3.615617                       0.270784   
min         1.000000            0.000000                       0.000000   
25%         1.000000            2.000000                       0.222000   
50%         1.000000            6.000000                       0.437000   
75%         2.000000            9.000000                       0.630000   
max       771.000000           11.000000                       1.000000   

       analysis_songs_mode  analysis_songs_mode_confidence  \
count        772661.000000                   772661.000000   
mean              0.637005                        0.489415   
std               0.480864                        0.165614   
min               0.000000                        0.000000   
25%               0.000000                        0.397000   
50%               1.000000                        0.480000   
75%               1.000000                        0.596000   
max               1.000000                        1.000000   

       analysis_songs_tempo  analysis_songs_time_signature  \
count         772661.000000                  772661.000000   
mean             126.560860                       3.730847   
std               31.789887                       1.021593   
min                0.000000                       0.000000   
25%              102.984000                       4.000000   
50%              121.884000                       4.000000   
75%              146.022000                       4.000000   
max              253.357000                       7.000000   

       analysis_songs_time_signature_confidence  musicbrainz_songs_year  \
count                             772661.000000           772661.000000   
mean                                   0.626652             1543.714245   
std                                    0.361508              839.540567   
min                                    0.000000                0.000000   
25%                                    0.358000             1977.000000   
50%                                    0.742000             2000.000000   
75%                                    0.994000             2006.000000   
max                                    1.000000             2010.000000   

        years_binned   tempo_binned  analysis_songs_duration  \
count  772661.000000  772661.000000            772661.000000   
mean        2.646629       4.783883                 0.158297   
std         1.620828       2.637524                 0.801748   
min         0.000000       0.000000                -2.151904   
25%         1.000000       3.000000                -0.228571   
50%         3.000000       5.000000                 0.052587   
75%         4.000000       7.000000                 0.464172   
max         4.000000       9.000000                14.586918   

       metadata_songs_artist_familiarity  metadata_songs_artist_hotttnesss  \
count                      772661.000000                     772661.000000   
mean                            1.137810                          1.828713   
std                             0.594691                          1.180343   
min                            -2.813196                         -2.960846   
25%                             0.847774                          1.077402   
50%                             1.298247                          1.707256   
75%                             1.572213                          2.589013   
max                             2.176964                          5.457244   

       metadata_songs_song_hotttnesss  
count                   772661.000000  
mean                         3.760755  
std                          2.317352  
min                         -4.007280  
25%                          2.414042  
50%                      

In [9]:
## Get our unique users
users = list(np.sort(standardized_df.user.unique()))
## Get our unique songs
songs = list(standardized_df.song.unique())
## Get counts of songs listened to
song_count = list(standardized_df['count'])

In [10]:
import scipy.sparse as sparse
from scipy.sparse.linalg import spsolve

## Get associated row indices
rows = pd.Categorical(standardized_df.user, categories=users).codes
## Get associated column indices
cols = pd.Categorical(standardized_df.song, categories=songs).codes
matrix = sparse.csr_matrix((song_count, (rows, cols)), shape=(len(users), len(songs)))

In [11]:
matrix

<418252x3675 sparse matrix of type '<class 'numpy.intc'>'
	with 772661 stored elements in Compressed Sparse Row format>

We have a total of 418,252 users and 3,675 songs, with 772,661 songs heard.

In [12]:
## Calculate matrix sparsity
# Obtain number of possible interactions in the matrix
matrix_size = matrix.shape[0]*matrix.shape[1]
# Obtain number of songs interacted with
num_songs_heard = len(matrix.nonzero()[0])
sparsity = 100*(1-(num_songs_heard/matrix_size))
sparsity

99.94973176669653

99.9% of the interaction matrix is sparse whereas the maximum that usually works for collaborative filtering is 99.5%. Therefore, we can attempt to build a matrix factorization model, but this may not perform well.

In [14]:
# Install a pip package in the current Jupyter kernel
import sys
!{sys.executable} -m pip install implicit

## Attempt 2 Using Towards Data Science

In [8]:
#content_matrix = sparse.csr_matrix((song_count, (rows, cols)), shape=(len(songs), len(users)))
import scipy.sparse as sparse
from scipy.sparse.linalg import spsolve
from scipy.sparse import csr_matrix

master_df['user']= master_df['user'].astype("category")
master_df['song']= master_df['song'].astype("category")
master_df['user_id']= master_df['user'].cat.codes
master_df['song_id']= master_df['song'].cat.codes

sparse_content_user = sparse.csr_matrix((master_df['count'].astype(float), (master_df['song_id'], master_df['user_id'])))
sparse_user_content = sparse.csr_matrix((master_df['count'].astype(float), (master_df['user_id'], master_df['song_id'])))
print (sparse_user_content)

  (0, 502)	2.0
  (0, 983)	1.0
  (0, 1907)	1.0
  (0, 2041)	1.0
  (0, 2842)	1.0
  (1, 87)	1.0
  (1, 291)	5.0
  (1, 688)	1.0
  (1, 826)	7.0
  (1, 958)	1.0
  (1, 2025)	5.0
  (1, 2698)	1.0
  (2, 838)	1.0
  (2, 989)	4.0
  (2, 1140)	2.0
  (2, 1824)	5.0
  (2, 2069)	1.0
  (2, 2236)	2.0
  (3, 87)	1.0
  (3, 769)	2.0
  (3, 1140)	9.0
  (3, 2236)	1.0
  (3, 2303)	8.0
  (3, 2566)	1.0
  (3, 2784)	1.0
  :	:
  (22836, 1467)	7.0
  (22836, 1577)	1.0
  (22836, 1723)	1.0
  (22836, 1741)	1.0
  (22836, 1909)	1.0
  (22836, 1951)	1.0
  (22836, 2645)	1.0
  (22836, 2905)	2.0
  (22837, 545)	1.0
  (22837, 733)	1.0
  (22837, 769)	1.0
  (22837, 1467)	1.0
  (22837, 1777)	1.0
  (22838, 87)	1.0
  (22838, 921)	1.0
  (22838, 978)	7.0
  (22838, 993)	1.0
  (22838, 2036)	1.0
  (22838, 2776)	2.0
  (22838, 2838)	1.0
  (22839, 409)	4.0
  (22839, 1538)	4.0
  (22839, 1951)	13.0
  (22839, 2575)	14.0
  (22839, 2604)	19.0


In [9]:
## Create a dataframe to look up user
user_lookup = master_df[['user', 'user_id']].drop_duplicates()
user_lookup.head()

user  user_id
3   724534729c9f5dc72a009269c2c225883e4775d2    10064
4   ee7aa84c164038c963cfd02a7e52a5598aa470c3    21294
5   1aa4fd215aadb160965110ed8a829745cde319eb     2322
8   cca2694a19b2474bc731574ebb8b10382214f865    18275
17  091d959cd1d21f270d9f6dc58b93ee78caa4ab96      790

In [10]:
## Create a dataframe to look up song
song_lookup = master_df[['song', 'song_id']].drop_duplicates()
song_lookup.head()

song  song_id
3     SOWEZSI12A81C21CE6     2567
771   SODCXXY12AB0187452      409
1364  SOWPAXV12A67ADA046     2612
5056  SOLXDDC12A6701FBFD     1486
7423  SONJBQX12A6D4F8382     1642

In [11]:
## Subset to data for model
model_df = master_df[['user_id', 'song_id', 'count']]
model_df.head()

user_id  song_id  count
3     10064     2567      1
4     21294     2567      2
5      2322     2567      1
8     18275     2567      2
17      790     2567     10

In [12]:
model_df.tail()

user_id  song_id  count
772636    19560     2523      1
772638    17062      271      5
772657    15017     1362      1
772658     9494     1543      1
772659     9494     2823      1

In [16]:
import implicit
## will need to tweak the number of latent factors by potentially identifying loss function stability
## will need to tweak regularization to tune parameters
## user cross validation to tune the optimum factors in model
## Reasonable solution usually occurs in 20 iterations or less
## Knowing the number of latent or hidden factors influencing the model is impossible. We should
## start at 5-10, then increase by 5 until performance of model stops improving
from sklearn.model_selection import StratifiedKFold
k_latent_factors = [35]
regularization_options = [0.01]
alpha_options = [15]
col_names=['Model', 'Hit Ratio', 'Mean_Avg_Precision', 'Mean_Avg_Recall']
x = model_df[['song_id', 'count']].to_numpy()
y = model_df[['user_id']].to_numpy()
model_evaluation_metrics_df = pd.DataFrame(columns=col_names)
n_splits = 5
skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=None)
skf.get_n_splits(model_df)
MAP = 0
for k in k_latent_factors:
    for i in regularization_options:
        for m in alpha_options:
            for train_index, test_index in skf.split(x, y):
                x_train, x_test = x[train_index], x[test_index]
                y_train, y_test = y[train_index], y[test_index]
                song_info = x_train[:,0]
                count_info = x_train[:,1]
                sparse_content_user = sparse.csr_matrix((count_info.astype(float), (song_info, y_train.flatten())))
                sparse_user_content = sparse.csr_matrix((count_info.astype(float),(y_train.flatten(), song_info)))
                model = implicit.als.AlternatingLeastSquares(factors=k, regularization=i, iterations=30)
                model.show_progress = False
                alpha = m
                data = (sparse_content_user*alpha).astype('double')
                model.fit(data)
                model_results = pd.DataFrame(data=x_test, columns=["song_id", 'count'])
                model_results['user_id']=pd.Series(y_test.flatten())
                model_results['predicted'] = model_results.apply(lambda row: model.recommend(row.user_id, sparse_user_content, N=10), axis=1)
                model_results['hits'] = model_results.apply(lambda row:[i for i in row['predicted'] if row['song_id'] in i], axis=1)
                model_results['hit'] = model_results.apply(lambda row: len(row['hits'])>0, axis=1)
                hits = len(model_results[model_results.hit==True])
                total = len(model_results)
                hit_ratio = hits/total
                user_actions = model_results.groupby('user_id').count()
                accurate_recommendations = model_results.groupby('user_id').sum()
                user_actions=user_actions[['song_id']]
                accurate_recommendations = accurate_recommendations[['hit']]
                metrics = user_actions.merge(accurate_recommendations, left_index=True, right_index=True)
                metrics['precision'] = metrics.hit/metrics.song_id
                metrics['recall'] = metrics.hit/10
                mean_average_precision = metrics.precision.mean()
                mean_average_recall = metrics.recall.mean()
                model_name = "factors:" + str(k) + " l2: " + str(i) + " alpha" + str(m)
                instance = {'Model': model_name, 'Hit Ratio': hit_ratio, 'Mean_Avg_Precision': mean_average_precision, 'Mean_Avg_Recall': mean_average_recall}
                model_evaluation_metrics_df = model_evaluation_metrics_df.append(instance, ignore_index=True)

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

  0%|          | 0/30 [00:00<?, ?it/s]

In [17]:
#model_evaluation_metrics_df = model_evaluation_metrics_df.groupby('Model').mean()
model_evaluation_metrics_df.sort_values(by=['Hit Ratio'], ascending=False)

Model  Hit Ratio  Mean_Avg_Precision  Mean_Avg_Recall
4  factors:35 l2: 0.01 alpha15   0.358578            0.361252         0.047430
0  factors:35 l2: 0.01 alpha15   0.358136            0.359517         0.047373
2  factors:35 l2: 0.01 alpha15   0.358115            0.360035         0.047369
3  factors:35 l2: 0.01 alpha15   0.357022            0.358306         0.047224
1  factors:35 l2: 0.01 alpha15   0.356017            0.358379         0.047093

Link to help explain confidence and alpha's impact on confidence:
https://stats.stackexchange.com/questions/155296/alternating-least-squares-als-why-two-different-kinds-of-setting-for-lambd

In [125]:
reldir = 'C:\\Users\\shefa\\OneDrive\\CS6220\\msd.recsys\\datasets\\'
model_evaluation_metrics_df.to_pickle(reldir + 'als_tuning_df2.csv', compression='zip')

In [29]:
## The paper used to develop ALS used alpha of 40
alpha = 40
data = (sparse_content_user*alpha).astype('double')
model.fit(data)

  0%|          | 0/20 [00:00<?, ?it/s]

In [30]:
## Obtain recommendations for a given user
user_id = 350627
recommended = model.recommend(user_id, sparse_user_content, N=5)
print(recommended)

[(693, 0.79169756), (1722, 0.71202666), (203, 0.70730686), (2730, 0.6374319), (2942, 0.63132167)]


In [31]:
## Get similar items

song_id=2918
n_similar = 3
similar = model.similar_items(song_id, n_similar)
print(similar)

[(2918, 0.99999994), (162, 0.8860025), (227, 0.8719206)]


https://campus.datacamp.com/courses/recommendation-engines-in-pyspark/what-if-you-dont-have-customer-ratings?ex=6

In [ ]:
#Use ROEM as evaluation metric